File: buffer/src/main/java/org/epdb/buffer/InMemoryBufferManager.java
package org.epdb.buffer;

import java.util.Collections;
import java.util.Map;

import org.epdb.buffer.dto.BufferFrame;
import org.epdb.storage.dto.Page;
import org.epdb.storage.manager.StorageManager;

public record InMemoryBufferManager(
        StorageManager storageManager,
        int bufferSize,
        Map<Long, BufferFrame> bufferFrames
) implements BufferManager {

    @Override
    public Page getPage(Long pageId) {
        if (this.bufferFrames.containsKey(pageId)) {
            var frame = this.bufferFrames.get(pageId);

            frame.pinCount++;
            return frame.page;
        }

        if (this.bufferFrames.size() >= this.bufferSize) {
            evictPage();
        }

        final var newPage = this.storageManager.readPage(pageId);
        final var frame = new BufferFrame(newPage);
        frame.pinCount++;
        this.bufferFrames.put(pageId, frame);

        return newPage;
    }

    @Override
    public void unpinPage(Long pageId, boolean isModified) {
        var bufferFrame = this.bufferFrames.get(pageId);

        if (bufferFrame == null) {
            throw new IllegalArgumentException("Page does not exist.");
        }

        if (bufferFrame.pinCount <= 0) {
            System.err.println("Illegal unpin action on page.");
        }

        bufferFrame.pinCount--;

        if (isModified) {
            bufferFrame.isDirty = true;
        }
    }

    @Override
    public void flushPage(BufferFrame frame) {
        if(frame.isDirty) {
            var page = frame.page;
            this.storageManager.writePage(page.getPageId(), page.getData());
            frame.isDirty = false;
        }
    }

    @Override
    public Page allocateNewPage(int tableId) {
        var newPageId = this.storageManager.allocatePage();
        return getPage(newPageId);
    }

    private void evictPage() {
        var victimKey = Collections.min(this.bufferFrames.keySet());
        flushPage(this.bufferFrames.get(victimKey));
        this.bufferFrames.remove(victimKey);
    }
}

File: buffer/src/main/kotlin/org/epdb/buffer/InMemoryBufferManager.kt
package org.epdb.buffer

import org.epdb.buffer.dto.BufferFrame
import org.epdb.storage.dto.Page
import org.epdb.storage.manager.InMemoryStorageManager
import org.epdb.storage.manager.StorageManager
import java.util.*

class InMemoryBufferManager(
    private val storageManager: StorageManager = InMemoryStorageManager(),
    private val bufferSize: Int = 10,
    private val bufferFrames: MutableMap<Long, BufferFrame> = mutableMapOf()
) : BufferManager {

    override fun getPage(pageId: Long): Page =
        bufferFrames.getOrPut(pageId) {
            if (bufferFrames.size >= bufferSize) {
                evictPage()
            }

            val newPage = storageManager.readPage(pageId)
            BufferFrame(newPage)
        }.also { it.pinCount++ }.page

    override fun unpinPage(pageId: Long, isModified: Boolean) {
        val bufferFrame = bufferFrames[pageId]?: run {
            System.err.println("Error: Page $pageId not found in buffer.")
            return
        }

        require (bufferFrame.pinCount > 0) {
            "Illegal unpin action on page ${pageId}. Pin count is already ${bufferFrame.pinCount}."
        }

        bufferFrame.apply {
            pinCount--
            isDirty = isDirty || isModified
        }
    }

    override fun flushPage(bufferFrame: BufferFrame) {
        if (bufferFrame.isDirty) {
            val page = bufferFrame.page
            storageManager.writePage(page.pageId, page.data)
            bufferFrame.isDirty = false
        }
    }

    override fun allocateNewPage(tableId: Int) = getPage(storageManager.allocatePage())

    private fun evictPage() {
        val victimKey = bufferFrames.keys.minOrNull()

        victimKey?.let { key ->
            val frameToEvict = bufferFrames.getValue(key)
            flushPage(frameToEvict)
            bufferFrames.remove(key)
        }
    }
}
